### ISSUE TITLE
mariadb connection refused

### ISSUE THREAD

**Issue by moeinrahimi** [docker-library/mariadb#299]
I have a docker-compose including nodejs and mariadb
```
version: '3'
services:
 web:
  # image: myapp
  build: .
  command: "npm start"
  volumes:
    - .:/app
    - /app/node_modules
  ports:
   - "8082:8082"
  depends_on:
    - "db"
  links:
    - db
 db:
  image: mariadb
  ports:
    - "3306:3306"
  volumes:
    - e:/data/dab:/var/lib/mysql
    - ./models/init-db.sql:/docker-entrypoint-initdb.d/init-db.sql:ro
  environment:
    MYSQL_ROOT_PASSWORD: moe2012@
    MYSQL_DATABASE: streamingnest
    MYSQL_USER: moein
    MYSQL_PASSWORD: moe2012@

```
I was getting connection refused and i was able to solve it by uncommenting bind-ip 0.0.0.0.
is this excepted behavior ? this is realy messy.
docker-compose up
```
db_1   | PLEASE REMEMBER TO SET A PASSWORD FOR THE MariaDB root USER !
db_1   | To do so, start the server, then issue the following commands:
db_1   | 
db_1   | '/usr/bin/mysqladmin' -u root password 'new-password'
db_1   | '/usr/bin/mysqladmin' -u root -h  password 'new-password'
db_1   |
db_1   | Alternatively you can run:
db_1   | '/usr/bin/mysql_secure_installation'
db_1   | 
db_1   | which will also give you the option of removing the test
db_1   | databases and anonymous user created by default.  This is
db_1   | strongly recommended for production servers.
db_1   | 
db_1   | See the MariaDB Knowledgebase at http://mariadb.com/kb or the
db_1   | MySQL manual for more instructions.
db_1   |
db_1   | Please report any problems at http://mariadb.org/jira
db_1   |
db_1   | The latest information about MariaDB is available at http://mariadb.org/.
db_1   | You can find additional information about the MySQL part at:
db_1   | http://dev.mysql.com
db_1   | Consider joining MariaDB's strong and vibrant community:
db_1   | https://mariadb.org/get-involved/
db_1   |
db_1   | 2020-04-14 20:09:30+00:00 [Note] [Entrypoint]: Database files initialized
db_1   | 2020-04-14 20:09:30+00:00 [Note] [Entrypoint]: Starting temporary server
db_1   | 2020-04-14 20:09:30+00:00 [Note] [Entrypoint]: Waiting for server startup
db_1   | 2020-04-14 20:09:31 0 [Note] mysqld (mysqld 10.4.12-MariaDB-1:10.4.12+maria~bionic-log) starting as process 126 ...
db_1   | mysqld: Can't create file '/dev/null.err' (errno: 13 "Permission denied")
db_1   | 2020-04-14 20:09:31 0 [Note] Using unique option prefix 'myisam_recover' is error-prone and can break in the future. Please use the full name 'myisam-recover-options' instead.
db_1   | 2020-04-14 20:09:31 0 [Note] InnoDB: Using Linux native AIO
db_1   | 2020-04-14 20:09:31 0 [Note] InnoDB: Mutexes and rw_locks use GCC atomic builtins
db_1   | 2020-04-14 20:09:31 0 [Note] InnoDB: Uses event mutexes
db_1   | 2020-04-14 20:09:31 0 [Note] InnoDB: Compressed tables use zlib 1.2.11
db_1   | 2020-04-14 20:09:31 0 [Note] InnoDB: Number of pools: 1
db_1   | 2020-04-14 20:09:31 0 [Note] InnoDB: Using SSE2 crc32 instructions
db_1   | 2020-04-14 20:09:31 0 [Note] mysqld: O_TMPFILE is not supported on /tmp (disabling future attempts)
db_1   | 2020-04-14 20:09:31 0 [Note] InnoDB: Initializing buffer pool, total size = 256M, instances = 1, chunk size = 128M
db_1   | 2020-04-14 20:09:31 0 [Note] InnoDB: Completed initialization of buffer pool
db_1   | 2020-04-14 20:09:31 0 [Note] InnoDB: If the mysqld execution user is authorized, page cleaner thread priority can be changed. See the man page of setpriority().   
db_1   | 2020-04-14 20:09:31 0 [Note] InnoDB: 128 out of 128 rollback segments are active.
db_1   | 2020-04-14 20:09:31 0 [Note] InnoDB: Creating shared tablespace for temporary tables
db_1   | 2020-04-14 20:09:31 0 [Note] InnoDB: Setting file './ibtmp1' size to 12 MB. Physically writing the file full; Please wait ...
db_1   | 2020-04-14 20:09:33 0 [Note] InnoDB: File './ibtmp1' size is now 12 MB.
db_1   | 2020-04-14 20:09:33 0 [Note] InnoDB: Waiting for purge to start
db_1   | 2020-04-14 20:09:33 0 [Note] InnoDB: 10.4.12 started; log sequence number 60972; transaction id 21
db_1   | 2020-04-14 20:09:33 0 [Note] Plugin 'FEEDBACK' is disabled.
db_1   | 2020-04-14 20:09:33 0 [Note] InnoDB: Loading buffer pool(s) from /var/lib/mysql/ib_buffer_pool
db_1   | 2020-04-14 20:09:33 0 [Note] InnoDB: Buffer pool(s) load completed at 200414 20:09:33
db_1   | 2020-04-14 20:09:33 0 [Warning] 'user' entry 'root@7839426991a8' ignored in --skip-name-resolve mode.
db_1   | 2020-04-14 20:09:33 0 [Warning] 'user' entry '@7839426991a8' ignored in --skip-name-resolve mode.
db_1   | 2020-04-14 20:09:33 0 [Warning] 'proxies_priv' entry '@% root@7839426991a8' ignored in --skip-name-resolve mode.
db_1   | 2020-04-14 20:09:33 0 [Note] Reading of all Master_info entries succeeded
db_1   | 2020-04-14 20:09:33 0 [Note] Added new Master_info '' to hash table
db_1   | 2020-04-14 20:09:33 0 [Note] mysqld: ready for connections.
db_1   | Version: '10.4.12-MariaDB-1:10.4.12+maria~bionic-log'  socket: '/var/run/mysqld/mysqld.sock'  port: 0  mariadb.org binary distribution
db_1   | 2020-04-14 20:09:34+00:00 [Note] [Entrypoint]: Temporary server started.
db_1   | Warning: Unable to load '/usr/share/zoneinfo/leap-seconds.list' as time zone. Skipping it.
db_1   | 2020-04-14 20:09:45 10 [Warning] 'proxies_priv' entry '@% root@7839426991a8' ignored in --skip-name-resolve mode.
db_1   | 2020-04-14 20:09:45+00:00 [Note] [Entrypoint]: Creating database streamingnest
db_1   | 2020-04-14 20:09:45+00:00 [Note] [Entrypoint]: Creating user moein
db_1   | 2020-04-14 20:09:45+00:00 [Note] [Entrypoint]: Giving user moein access to schema streamingnest
db_1   | 2020-04-14 20:09:45 14 [Warning] 'proxies_priv' entry '@% root@7839426991a8' ignored in --skip-name-resolve mode.
db_1   |
db_1   | 2020-04-14 20:09:45+00:00 [Note] [Entrypoint]: /usr/local/bin/docker-entrypoint.sh: running /docker-entrypoint-initdb.d/init-db.sql
db_1   | ERROR: Can't initialize batch_readline - may be the input source is a directory or a block device.
backend_db_1 exited with code 1
```
I event include a init.sql
```

create database streamingnest;

CREATE USER 'moein' IDENTIFIED BY 'moe2012@';
GRANT USAGE ON *.* TO 'moein'@'%' IDENTIFIED BY 'moe2012@';
GRANT ALL privileges ON `streamingnest`.* TO 'moein1'@%;
FLUSH PRIVILEGES;
```

update: bind-address to 0.0.0.0 not working anymore, not sure whats wrong.

**Comment by wglambert** [comment]
Can you give all the commands you ran and any relevant logs for reproducing the issue

**Comment by moeinrahimi** [comment]
@wglambert I updated my issue with more info, also I tried docker-compose down and up including deleting volume folder.

**Comment by wglambert** [comment]
>ERROR: Can't initialize batch_readline - may be the input source is a directory or a block device.

Is that file mounted correctly?
There's one issue on that error https://github.com/docker-library/mysql/issues/265

I don't see the `connection refused` error but I can connect fine from the host or another container (to the host on the forwarded port)

```console
$ docker run -d --rm -e MYSQL_ROOT_PASSWORD=pass --name mariadb -p 3306:3306 mariadb
01e221a3724ff6699ff9a5b35446f133d31ad5f06220a132c7f715aa6eae5ece

$ docker ps
CONTAINER ID        IMAGE               COMMAND                  CREATED             STATUS              PORTS                    NAMES
01e221a3724f        mariadb             "docker-entrypoint.s…"   3 minutes ago       Up 3 minutes        0.0.0.0:3306->3306/tcp   mariadb

$ mysql -h 127.0.0.1 -uroot -ppass
Welcome to the MariaDB monitor.  Commands end with ; or \g.
Your MariaDB connection id is 8
Server version: 10.4.12-MariaDB-1:10.4.12+maria~bionic mariadb.org binary distribution

Copyright (c) 2000, 2018, Oracle, MariaDB Corporation Ab and others.

Type 'help;' or '\h' for help. Type '\c' to clear the current input statement.

MariaDB [(none)]> exit
Bye
```

From container -> host -> Mariadb container
```console
$ docker run -it --rm mysql mysql -uroot -ppass -h 172.17.0.1
mysql: [Warning] Using a password on the command line interface can be insecure.
Welcome to the MySQL monitor.  Commands end with ; or \g.
Your MySQL connection id is 9
Server version: 5.5.5-10.4.12-MariaDB-1:10.4.12+maria~bionic mariadb.org binary distribution

Copyright (c) 2000, 2020, Oracle and/or its affiliates. All rights reserved.

Oracle is a registered trademark of Oracle Corporation and/or its
affiliates. Other names may be trademarks of their respective
owners.

Type 'help;' or '\h' for help. Type '\c' to clear the current input statement.

mysql>
```

**Comment by moeinrahimi** [comment]
I commented init script to see the result.
```
web_1  | Sequelize issue:
web_1  | err name :SequelizeConnectionRefusedError
web_1  | err message :  connect ECONNREFUSED 172.22.0.2:3306
db_1   | 
db_1   |
db_1   | PLEASE REMEMBER TO SET A PASSWORD FOR THE MariaDB root USER !
db_1   | To do so, start the server, then issue the following commands:
db_1   | 
db_1   | '/usr/bin/mysqladmin' -u root password 'new-password'
db_1   | '/usr/bin/mysqladmin' -u root -h  password 'new-password'
db_1   |
db_1   | Alternatively you can run:
db_1   | '/usr/bin/mysql_secure_installation'
db_1   |
db_1   | which will also give you the option of removing the test
db_1   | databases and anonymous user created by default.  This is
db_1   | strongly recommended for production servers.
db_1   |
db_1   | See the MariaDB Knowledgebase at http://mariadb.com/kb or the
db_1   | MySQL manual for more instructions.
db_1   | 
db_1   | Please report any problems at http://mariadb.org/jira
db_1   |
db_1   | The latest information about MariaDB is available at http://mariadb.org/.
db_1   | You can find additional information about the MySQL part at:
db_1   | http://dev.mysql.com
db_1   | Consider joining MariaDB's strong and vibrant community:
db_1   | https://mariadb.org/get-involved/
db_1   | 
db_1   | 2020-04-15 11:34:54+00:00 [Note] [Entrypoint]: Database files initialized
db_1   | 2020-04-15 11:34:54+00:00 [Note] [Entrypoint]: Starting temporary server
db_1   | 2020-04-15 11:34:54+00:00 [Note] [Entrypoint]: Waiting for server startup
db_1   | 2020-04-15 11:34:54 0 [Note] mysqld (mysqld 10.4.12-MariaDB-1:10.4.12+maria~bionic) starting as process 126 ...
db_1   | 2020-04-15 11:34:54 0 [Note] InnoDB: Using Linux native AIO
db_1   | 2020-04-15 11:34:55 0 [Note] InnoDB: Mutexes and rw_locks use GCC atomic builtins
db_1   | 2020-04-15 11:34:55 0 [Note] InnoDB: Uses event mutexes
db_1   | 2020-04-15 11:34:55 0 [Note] InnoDB: Compressed tables use zlib 1.2.11
db_1   | 2020-04-15 11:34:55 0 [Note] InnoDB: Number of pools: 1
db_1   | 2020-04-15 11:34:55 0 [Note] InnoDB: Using SSE2 crc32 instructions
db_1   | 2020-04-15 11:34:55 0 [Note] mysqld: O_TMPFILE is not supported on /tmp (disabling future attempts)
db_1   | 2020-04-15 11:34:55 0 [Note] InnoDB: Initializing buffer pool, total size = 256M, instances = 1, chunk size = 128M
db_1   | 2020-04-15 11:34:55 0 [Note] InnoDB: Completed initialization of buffer pool
db_1   | 2020-04-15 11:34:55 0 [Note] InnoDB: If the mysqld execution user is authorized, page cleaner thread priority can be changed. See the man page of setpriority().   
db_1   | 2020-04-15 11:34:55 0 [Note] InnoDB: 128 out of 128 rollback segments are active.
db_1   | 2020-04-15 11:34:55 0 [Note] InnoDB: Creating shared tablespace for temporary tables
db_1   | 2020-04-15 11:34:55 0 [Note] InnoDB: Setting file './ibtmp1' size to 12 MB. Physically writing the file full; Please wait ...
db_1   | 2020-04-15 11:34:56 0 [Note] InnoDB: File './ibtmp1' size is now 12 MB.
db_1   | 2020-04-15 11:34:56 0 [Note] InnoDB: 10.4.12 started; log sequence number 60972; transaction id 21
db_1   | 2020-04-15 11:34:56 0 [Note] Plugin 'FEEDBACK' is disabled.
db_1   | 2020-04-15 11:34:56 0 [Note] InnoDB: Loading buffer pool(s) from /var/lib/mysql/ib_buffer_pool
db_1   | 2020-04-15 11:34:56 0 [Note] InnoDB: Buffer pool(s) load completed at 200415 11:34:56
db_1   | 2020-04-15 11:34:56 0 [Warning] 'user' entry 'root@18e2bd1fb54b' ignored in --skip-name-resolve mode.
db_1   | 2020-04-15 11:34:56 0 [Warning] 'user' entry '@18e2bd1fb54b' ignored in --skip-name-resolve mode.
db_1   | 2020-04-15 11:34:56 0 [Warning] 'proxies_priv' entry '@% root@18e2bd1fb54b' ignored in --skip-name-resolve mode.
db_1   | 2020-04-15 11:34:56 0 [Note] Reading of all Master_info entries succeeded
db_1   | 2020-04-15 11:34:56 0 [Note] Added new Master_info '' to hash table
db_1   | 2020-04-15 11:34:56 0 [Note] mysqld: ready for connections.
db_1   | Version: '10.4.12-MariaDB-1:10.4.12+maria~bionic'  socket: '/var/run/mysqld/mysqld.sock'  port: 0  mariadb.org binary distribution
db_1   | 2020-04-15 11:34:57+00:00 [Note] [Entrypoint]: Temporary server started.
db_1   | Warning: Unable to load '/usr/share/zoneinfo/leap-seconds.list' as time zone. Skipping it.
db_1   | 2020-04-15 11:35:07 10 [Warning] 'proxies_priv' entry '@% root@18e2bd1fb54b' ignored in --skip-name-resolve mode.
db_1   | 2020-04-15 11:35:07+00:00 [Note] [Entrypoint]: Creating database streamingnest
db_1   | 2020-04-15 11:35:07+00:00 [Note] [Entrypoint]: Creating user moein
db_1   | 2020-04-15 11:35:07+00:00 [Note] [Entrypoint]: Giving user moein access to schema streamingnest
db_1   | 2020-04-15 11:35:07 14 [Warning] 'proxies_priv' entry '@% root@18e2bd1fb54b' ignored in --skip-name-resolve mode.
db_1   |
db_1   | 2020-04-15 11:35:07+00:00 [Note] [Entrypoint]: Stopping temporary server
db_1   | 2020-04-15 11:35:08 0 [Note] mysqld (initiated by: root[root] @ localhost []): Normal shutdown
db_1   | 2020-04-15 11:35:08 0 [Note] Event Scheduler: Purging the queue. 0 events
db_1   | 2020-04-15 11:35:08 0 [Note] InnoDB: FTS optimize thread exiting.
db_1   | 2020-04-15 11:35:08 0 [Note] InnoDB: Starting shutdown...
db_1   | 2020-04-15 11:35:08 0 [Note] InnoDB: Dumping buffer pool(s) to /var/lib/mysql/ib_buffer_pool
db_1   | 2020-04-15 11:35:08 0 [Note] InnoDB: Buffer pool(s) dump completed at 200415 11:35:08
db_1   | 2020-04-15 11:35:09 0 [Note] InnoDB: Shutdown completed; log sequence number 60981; transaction id 24
db_1   | 2020-04-15 11:35:09 0 [Note] InnoDB: Removed temporary tablespace data file: "ibtmp1"
db_1   | 2020-04-15 11:35:09 0 [Note] mysqld: Shutdown complete
db_1   |
db_1   | 2020-04-15 11:35:10+00:00 [Note] [Entrypoint]: Temporary server stopped
db_1   |
db_1   | 2020-04-15 11:35:10+00:00 [Note] [Entrypoint]: MySQL init process done. Ready for start up.
db_1   |
db_1   | 2020-04-15 11:35:10 0 [Note] mysqld (mysqld 10.4.12-MariaDB-1:10.4.12+maria~bionic) starting as process 1 ...
db_1   | 2020-04-15 11:35:10 0 [Note] InnoDB: Using Linux native AIO
db_1   | 2020-04-15 11:35:10 0 [Note] InnoDB: Mutexes and rw_locks use GCC atomic builtins
db_1   | 2020-04-15 11:35:10 0 [Note] InnoDB: Uses event mutexes
db_1   | 2020-04-15 11:35:10 0 [Note] InnoDB: Compressed tables use zlib 1.2.11
db_1   | 2020-04-15 11:35:10 0 [Note] InnoDB: Number of pools: 1
db_1   | 2020-04-15 11:35:10 0 [Note] InnoDB: Using SSE2 crc32 instructions
db_1   | 2020-04-15 11:35:10 0 [Note] mysqld: O_TMPFILE is not supported on /tmp (disabling future attempts)
db_1   | 2020-04-15 11:35:10 0 [Note] InnoDB: Initializing buffer pool, total size = 256M, instances = 1, chunk size = 128M
db_1   | 2020-04-15 11:35:10 0 [Note] InnoDB: Completed initialization of buffer pool
db_1   | 2020-04-15 11:35:10 0 [Note] InnoDB: If the mysqld execution user is authorized, page cleaner thread priority can be changed. See the man page of setpriority().   
db_1   | 2020-04-15 11:35:10 0 [Note] InnoDB: 128 out of 128 rollback segments are active.
db_1   | 2020-04-15 11:35:10 0 [Note] InnoDB: Creating shared tablespace for temporary tables
db_1   | 2020-04-15 11:35:10 0 [Note] InnoDB: Setting file './ibtmp1' size to 12 MB. Physically writing the file full; Please wait ...
db_1   | 2020-04-15 11:35:12 0 [Note] InnoDB: File './ibtmp1' size is now 12 MB.
db_1   | 2020-04-15 11:35:12 0 [Note] InnoDB: Waiting for purge to start
db_1   | 2020-04-15 11:35:12 0 [Note] InnoDB: 10.4.12 started; log sequence number 60981; transaction id 21
db_1   | 2020-04-15 11:35:12 0 [Note] Plugin 'FEEDBACK' is disabled.
db_1   | 2020-04-15 11:35:12 0 [Note] InnoDB: Loading buffer pool(s) from /var/lib/mysql/ib_buffer_pool
db_1   | 2020-04-15 11:35:12 0 [Note] InnoDB: Buffer pool(s) load completed at 200415 11:35:12
db_1   | 2020-04-15 11:35:12 0 [Note] Server socket created on IP: '::'.
db_1   | 2020-04-15 11:35:12 0 [Warning] 'proxies_priv' entry '@% root@18e2bd1fb54b' ignored in --skip-name-resolve mode.
db_1   | 2020-04-15 11:35:12 0 [Note] Reading of all Master_info entries succeeded
db_1   | 2020-04-15 11:35:12 0 [Note] Added new Master_info '' to hash table
db_1   | 2020-04-15 11:35:12 0 [Note] mysqld: ready for connections.
db_1   | Version: '10.4.12-MariaDB-1:10.4.12+maria~bionic'  socket: '/var/run/mysqld/mysqld.sock'  port: 3306  mariadb.org binary distribution
```

**Comment by wglambert** [comment]
Your `web_1` container is trying to connect immediately, mariadb begins initializing at `34:54` and finishes at `35:12` so if you try the connection again after say 20-30 seconds then it should work

**Comment by moeinrahimi** [comment]
so how am I suppose to wait for it? the depends option in docker-compose won't do the trick ?

**Comment by yosifkit** [comment]
> the depends option in docker-compose won't do the trick ?

Nope, that just controls start order, not "wait until ready". That would need to be either in your service (to retry connection a number of times) or a script to orchestrate your compose file to start the db and wait for it to come up before starting your other service.
